### PR TITLE
[Platform] Introduce ProviderFactory and ProviderConfigFactory

### LINKED ISSUE #402: Implement Platform DSN for AI Provider Configuration
## Problem
Currently, AI platform configuration requires multiple parameters (API key, endpoint, version, etc.) to be set separately. Symfony already uses DSN patterns successfully for mailers and notifiers, providing a clean and standardized way to configure external services.
## Proposed Solution
Implement a DSN (Data Source Name) format for AI platform configuration, following Symfony's established patterns:
```
scheme://API_KEY@HOST[:PORT][/PATH]?option1=value1&option2=value2
```
### Basic Usage
```php
// Anthropic
'anthropic://sk-ant-api03-xxx@api.anthropic.com?version=2025-01-01'
// OpenAI
'openai://sk-xxx@api.openai.com/v1?model=gpt-4'
// Azure OpenAI
'azure://API_KEY@myinstance.openai.azure.com?api-version=2024-02-01&deployment=gpt-4'
// Custom/Self-hosted
'ollama://localhost:11434?model=llama2'
```
### Environment Variables
```env
# .env file
AI_PLATFORM_DSN=anthropic://sk-ant-api03-xxx@api.anthropic.com?version=2025-01-01
```
### Configuration
```yaml
# config/packages/ai.yaml
ai:
    platform:
        dsn: '%env(AI_PLATFORM_DSN)%'
```
### Bundle Configuration Example
```yaml
# config/packages/ai.yaml
ai:
    agents:
        my_assistant:
            platform: 'anthropic://sk-ant-api03-xxx@api.anthropic.com?version=2025-01-01&model=claude-3-opus'
        
        code_reviewer:
            platform: 'openai://sk-xxx@api.openai.com/v1?model=gpt-4&temperature=0.2'
        
        local_agent:
            platform: 'ollama://localhost:11434?model=llama2'
```

### Platform Configuration Simplification
Current platform configuration:
```yaml
ai:
    platform:
        openai:
            api_key: '%env(OPENAI_API_KEY)%'
        azure:
            default:
                base_url: '%env(AZURE_OPENAI_BASEURL)%'
                deployment: '%env(AZURE_OPENAI_GPT)%'
                api_key: '%env(AZURE_OPENAI_API_KEY)%'
                api_version: '%env(AZURE_OPENAI_VERSION)%'
            embeddings:
                base_url: '%env(AZURE_OPENAI_EMBEDDINGS_BASEURL)%'
                deployment: '%env(AZURE_OPENAI_EMBEDDINGS)%'
                api_key: '%env(AZURE_OPENAI_EMBEDDINGS_API_KEY)%'
                api_version: '%env(AZURE_OPENAI_VERSION)%'
```
With DSN:
```yaml
ai:
    platform:
        openai: '%env(OPENAI_DSN)%'  # openai://sk-xxx@api.openai.com/v1
        azure:
            default: '%env(AZURE_DSN)%'  # azure://API_KEY@instance.openai.azure.com?deployment=gpt-4&api-version=2024-02-01
            embeddings: '%env(AZURE_EMBEDDINGS_DSN)%'  # azure://API_KEY@instance.openai.azure.com?deployment=embeddings&api-version=2024-02-01
```
## DSN Components
- **Scheme**: Provider name (`anthropic`, `openai`, `azure`, `ollama`, etc.)
- **User**: API key or authentication token
- **Host**: API endpoint URL
- **Port**: Optional port number (default varies by provider)
- **Path**: Optional API path
- **Query Parameters**: 
  - `version`: API version
  - `model`: Default model
  - `timeout`: Request timeout
  - `max_retries`: Retry configuration
  - Provider-specific options
## Benefits
- **Consistency**: Follows Symfony's established DSN patterns (mailer, notifier)
- **Simplicity**: Single configuration string contains all connection details
- **Security**: Easy to use environment variables for sensitive data
- **Flexibility**: Query parameters allow provider-specific configurations
- **Familiarity**: Developers already know this pattern from other Symfony components
## Implementation Notes
- Parse DSN using Symfony's existing DSN parsing utilities (across different components)
- Support only full DSN and no individual parameters in configuration anymore (NO backward compatibility)
- Validate scheme against supported providers
- Handle provider-specific requirements in dedicated factories
## References
- [Symfony Mailer DSN](https://symfony.com/doc/current/mailer.html#transport-setup)
- [Symfony Notifier DSN](https://symfony.com/doc/current/notifier.html#channel-configuration)

### PR DESCRIPTION
**Pull request by qbejs**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes 
| Docs?         | yes 
| Issues        | Fix #402
| License       | MIT

This PR implements **RFC #402** by introducing a `ProviderFactory` and `ProviderConfigFactory`
to centralize the creation of AI provider platforms (OpenAI, Azure OpenAI, Azure Meta) from DSNs.

### What’s new

- **`ProviderConfigFactory::fromDsn()`**  
  Parses DSNs into a strongly typed `ProviderConfig` (provider, baseUri, apiKey, options, headers).  
  Supports Azure family selection via `engine` (e.g. `openai`, `meta`) and common options like `deployment`, `version`, `timeout`, `proxy`, `verify_peer`, as well as `headers[...]`.

- **`ProviderFactory::fromDsn()`**  
  Instantiates the correct bridge `PlatformFactory` (OpenAi, Azure\OpenAi, Azure\Meta) using the parsed config.  
  Sets proper authentication headers per provider (`Authorization: Bearer …` or `api-key` for Azure), and propagates `base_uri`, `options`, and `headers` to the bridge.

- **Unit tests (positive & negative)**  
  Test coverage for both factories, including edge-cases for DSN parsing (e.g. no host → defaults, Azure validation).  
  A **dedicated PHPUnit configuration** `phpunit.provider-factory.xml` runs tests that depend on **fake bridge factories** (see rationale below).  
  A standard suite continues to run with the default `phpunit.xml(.dist)` without interference.

- **Changelog**  
  Entry added in `src/platform/CHANGELOG.md`.

### Why

This provides a unified, extensible mechanism to construct provider platforms from DSNs.
It makes configuration easier, allows future providers to be added consistently,
and aligns with the design goals outlined in **RFC #402**.

---

### Why a dedicated PHPUnit config for some tests?

The existing bridges expose **static** factory methods (e.g. `Bridge\*\PlatformFactory::create(...)`).  
Static factories are not mock-friendly: traditional PHPUnit doubles cannot intercept static calls and signatures differ per bridge.  
To test `ProviderFactory::fromDsn()` deterministically **without** coupling the suite to real providers, we load **fake bridge factories** under the same FQCN via a **bootstrap** that prepends a PSR-4 mapping for the `Bridge\` namespace.

- The **special suite** (`phpunit.provider-factory.xml`) uses `tests/bootstrap_fake_bridges.php` to ensure fake classes are loaded **before** the real ones.
- The main suite keeps using the default bootstrap, so **real bridge tests remain untouched**.
- These special tests are grouped (e.g. `@group pf`) and excluded from the main suite to avoid accidental mixing.

This split lets us:
- keep real bridge tests pristine,
- test the orchestration logic in `ProviderFactory` in isolation,
- avoid reflective heuristics around static method signatures.

---

### Proposed follow-up Improvement (separate PR/RFC)

To remove the need for a separate PHPUnit config and fakes in the future, I propose introducing a thin, optional abstraction layer:

1. **`BridgePlatformFactoryInterface`** (small contract)  
   ```php
   interface BridgePlatformFactoryInterface {
       public function create(string $apiKey, HttpClientInterface $http, array $contract): object;
       public static function providerKey(): string;   // e.g. "openai", "azure"
       public static function engine(): ?string;       // e.g. "openai" | "meta" for Azure, null otherwise
   }

Adapters per existing bridge would simply delegate to the current static PlatformFactory::create(...).

2. **`BridgeFactoryResolverInterface`** + default implementation
A registry that maps (providerKey, engine) → the corresponding adapter (injected via DI).
ProviderFactory::fromDsn() would ask the resolver for the correct adapter, then call ->create(...).

### CODE REVIEW

File: src/platform/CHANGELOG.md [hunk 1]
```diff
@@ -61,3 +61,7 @@ CHANGELOG
  * Add tool calling support for Ollama platform
  * Allow beta feature flags to be passed into Anthropic model options
  * Add Ollama streaming output support
+
+## [Unreleased]
+
+- feature #402 [AI] Introduced `ProviderFactory` and `ProviderConfigFactory` to create AI provider platforms from DSNs.
```
Comment (chr-hertel) [inline src/platform/CHANGELOG.md]: nothing is "released" with a tag:
```suggestion

## [Unreleased]

- Introduced `ProviderFactory` and `ProviderConfigFactory` to create AI provider platforms from DSNs.
```

File: src/platform/src/Transport/Dsn.php [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\AI\Platform\Transport;
+
+use Symfony\AI\Platform\Exception\InvalidArgumentException;
+
+final class Dsn
```
Comment (OskarStark) [inline src/platform/src/Transport/Dsn.php:16]: Where does this code com from? We have several `DSN` classes in `symfony/symfony` which should be used. I propose to copy it from `symonfy/notifer` including tests.

File: src/platform/src/Transport/Dsn.php [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    /** @return array<string,mixed> */
+    public function getQuery(): array
+    {
+        return $this->query;
+    }
+
+    public function getProvider(): string
+    {
+        $scheme = strtolower($this->scheme);
+        if (str_starts_with($scheme, 'ai+')) {
```
Comment (OskarStark) [inline src/platform/src/Transport/Dsn.php:134]: no need for a prefix here

File: src/platform/tests/Factory/FakeBridges/Azure/Meta/PlatformFactory.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\AI\Platform\Bridge\Azure\Meta;
+
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+final class PlatformFactory
```
Comment (OskarStark) [inline src/platform/tests/Factory/FakeBridges/Azure/Meta/PlatformFactory.php:16]: these classes look weird

**Comment by OskarStark** [comment]
Glad to have you here @qbejs 👋 😄

File: src/platform/CHANGELOG.md [hunk 5]
```diff
@@ -61,3 +61,7 @@ CHANGELOG
  * Add tool calling support for Ollama platform
  * Allow beta feature flags to be passed into Anthropic model options
  * Add Ollama streaming output support
+
+## [Unreleased]
+
+- Introduced `ProviderFactory` and `ProviderConfigFactory` to create AI provider platforms from DSNs.
```
Comment (OskarStark) [inline src/platform/CHANGELOG.md:67]:
```suggestion
- `ProviderFactory` and `ProviderConfigFactory` to create AI provider platforms from DSNs
```

File: src/platform/src/Factory/ProviderConfigFactory.php [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    private static function defaultHostOrFail(string $provider): string
+    {
+        return match ($provider) {
+            'openai' => 'api.openai.com',
+            'anthropic' => 'api.anthropic.com',
+            'gemini' => 'generativelanguage.googleapis.com',
+            'vertex' => 'us-central1-aiplatform.googleapis.com',
+            'ollama' => 'localhost',
+            'azure' => throw new InvalidArgumentException('Azure DSN must specify host (e.g. "<resource>.openai.azure.com").'),
+            default => throw new InvalidArgumentException(\sprintf('Unknown AI provider "%s".', $provider)),
```
Comment (OskarStark) [inline src/platform/src/Factory/ProviderConfigFactory.php:132]: Those should not be defined in the `DSN` class, but rather (like symfony/notfier does) in the AIBundle.php, ofc SfNotfier does this in FrameworkBundle

File: src/platform/src/Factory/ProviderFactory.php [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    public function fromDsn(string $dsn): object
+    {
+        $config = ProviderConfigFactory::fromDsn($dsn);
+        $providerKey = strtolower($config->provider);
+
+        if ('azure' === $providerKey) {
+            $engine = strtolower($config->options['engine'] ?? 'openai');
+            $factoryFqcn = match ($engine) {
```
Comment (OskarStark) [inline src/platform/src/Factory/ProviderFactory.php:31]: Lets go with dedicated PlatformFactory classes in each bridge (like the bridges in symfony/symfony)

**Comment by chr-hertel** [comment]
I'd prefer not to introduce the new term Provider here, but stick to Platform

**Comment by OskarStark** [comment]
Closing here for now, thanks for proposing
